### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by akshaisarma**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+import static org.apache.http.entity.ContentType.APPLICATION_JSON;
+
+/**
+ * This class publishes a {@link MetricEvent} to a provided URL. It can retry publishes.
+ */
+@Slf4j
+public class HTTPMetricEventPublisher implements MetricEventPublisher {
+    private final String url;
+    private final String group;
```
Comment (0aix) [inline src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java]: It looks like "group" is unused (?)

File: src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    private final int retries;
+    private final int retryIntervalMS;
+    private final Map<String, String> dimensions;
```
Comment (0aix) [inline src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java]: Is the name "dimensions" appropriate outside of yamas?

File: src/main/java/com/yahoo/bullet/common/metrics/MetricEvent.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+package com.yahoo.bullet.common.metrics;
+
+import com.yahoo.bullet.result.JSONFormatter;
+import lombok.Getter;
+
+import java.util.Map;
+
+/**
+ * Simpler wrapper that represents a metric event being published.
```
Comment (0aix) [inline src/main/java/com/yahoo/bullet/common/metrics/MetricEvent.java]: Simple*

File: src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java (hunk 2 above)
Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java]: Yea it's just tags describing the thing in more detail

File: src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            return false;
+        } catch (Exception e) {
+            request.abort();
+            log.error("Unable to publish request", e);
+            return false;
+        }
+    }
+
+    private boolean shouldRetry(boolean status, int count, int retries, int interval) {
+        boolean shouldNotSleep = status || count >= retries;
```
[the remark this answers is not in the document]
Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java]: But it'd be do { } while(!shouldNotRetry(...));

File: src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java (hunk 1 above)
Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java]: If you publish a MetricEvent, `getGroup` is called in convert in MetricEventPublisher

**Comment by akshaisarma** [comment]
Will add a BulletConfig taking constructor

File: src/main/java/com/yahoo/bullet/common/Validator.java [hunk 5]
```diff
@@ -751,6 +761,29 @@ public static boolean isClassName(Object value) {
...
+     * @param keyType The class of the key of the map to check for.
+     * @param valueType The class of the value of the map to check for.
+     * @param <K> The type of the key in the map.
+     * @param <V> The type of the value in the map.
+     * @return A Predicate that checks tor see if the value was a non-empty map of the given types.
+     */
+    @SuppressWarnings("unchecked")
+    public static <K, V> Predicate<Object> isMapOfType(Class<K> keyType, Class<V> valueType) {
+        return value -> {
+            if (!isNonEmptyMap(value)) {
```
Comment (NathanSpeidel) [inline src/main/java/com/yahoo/bullet/common/Validator.java]: Why do we insist the map is not empty?

Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/common/Validator.java]: Generics type erasure. Need key, values to check their types

File: src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+    @Override
+    public CompletableFuture<Boolean> publish(MetricEvent payload) {
+        String json = payload.asJSON();
+        log.debug("Publishing metric {}", json);
+        HttpUriRequest post = getPost(json);
+        return CompletableFuture.supplyAsync(() -> submitWithRetry(post, retries))
+                                .thenApply(HTTPMetricEventPublisher::onHTTPResult)
+                                .exceptionally(HTTPMetricEventPublisher::onHTTPFail);
+
```
Comment (NathanSpeidel) [inline src/main/java/com/yahoo/bullet/common/metrics/HTTPMetricEventPublisher.java]: Extra line?

**Pull request merged**